src/backend/x11/window_inner.rs

```diff
@@ -0,0 +1,238 @@
+use std::sync::{
+    atomic::{AtomicBool, Ordering},
+    Arc, Mutex, Weak,
+};
+use x11rb::{
+    connection::Connection,
+    protocol::{
+        xfixes::ConnectionExt,
+        xproto::{
+            self as x11, AtomEnum, ConnectionExt as _, CreateGCAux, CreateWindowAux, Depth, EventMask,
+            PropMode, Screen, UnmapNotifyEvent, WindowClass,
+        },
+    },
+    rust_connection::RustConnection,
+    wrapper::ConnectionExt as _,
+};
+
+impl From<Arc<WindowInner>> for Window {
+    fn from(inner: Arc<WindowInner>) -> Self {
+        Window(Arc::downgrade(&inner))
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct WindowInner {
+    // TODO: Consider future x11rb WindowWrapper
+    pub connection: Weak<RustConnection>,
+    pub id: x11::Window,
+    root: x11::Window,
+    pub atoms: Atoms,
+    pub size: Mutex<Size<u16, Logical>>,
+    /// Maintains the current cursor visibility state.
+    ///
+    /// Required since we cannot enable an already enabled cursor in xfixes.
+    cursor_visible: Arc<AtomicBool>,
+    pub depth: Depth,
+    pub gc: x11::Gcontext,
+}
+
+impl WindowInner {
+    pub fn new(
+        connection: Weak<RustConnection>,
+        screen: &Screen,
+        properties: WindowProperties<'_>,
+        atoms: Atoms,
+        depth: Depth,
+        visual_id: u32,
+        colormap: u32,
+    ) -> Result<WindowInner, X11Error> {
+        let weak = connection;
+        let connection = weak.upgrade().unwrap();
+
+        // Generate the xid for the window
+        let window = connection.generate_id()?;
+
+        // The event mask never include `EventMask::RESIZE_REDIRECT`.
+        //
+        // The reason is twofold:
+        // - We are not a window manager
+        // - Makes our window impossible to resize.
+        //
+        // On the resizing aspect, KWin and some other WMs would allow resizing, but those
+        // compositors rely on putting this window in another window for drawing decorations,
+        // so visibly in KWin it would look like using the RESIZE_REDIRECT event mask would work,
+        // but a tiling window manager would be sad and the tiling window manager devs mad because
+        // this window would refuse to listen to the tiling WM.
+        //
+        // For resizing we use ConfigureNotify events from the STRUCTURE_NOTIFY event mask.
+
+        let window_aux = CreateWindowAux::new()
+            .event_mask(
+                EventMask::EXPOSURE // Be told when the window is exposed
+            | EventMask::STRUCTURE_NOTIFY
+            | EventMask::KEY_PRESS // Key press and release
+            | EventMask::KEY_RELEASE
+            | EventMask::BUTTON_PRESS // Mouse button press and release
+            | EventMask::BUTTON_RELEASE
+            | EventMask::POINTER_MOTION // Mouse movement
+            | EventMask::NO_EVENT,
+            )
+            // Border pixel and color map need to be set if our depth may differ from the root depth.
+            .border_pixel(0)
```

Comment:
Instead of hardcoding a zero, I would suggest `connection.setup().black_pixel`. That will most likely still be the value 0, but... yeah.